File: typescript/src/core/modules/pcustomer-management/functions/update-customer.ts
import { AppError, isStandardError } from "../../../error";
import { PCustomerDAO } from "../data-model/dao";
import { initializeInternalContext } from "../../../context/internal-context";

// Import types
import type { RuntimeContext } from "../../../context/runtime-context";
import type { InternalContext } from "../../../context/internal-context";
import type { TPCustomer } from "../data-model/type";

/**
 * Cập nhật thông tin của một khách hàng trong hệ thống.
 *
 * @param ctx - runtime context.
 *
 * @returns
 */
export async function updateCustomer(ctx: RuntimeContext) {
  try {
    const body = await ctx.getBody<Partial<TPCustomer>>();
    const params = await ctx.getParams<{ id: string }>();

    const pcustomerDao = new PCustomerDAO();
    const internalCtx = initializeInternalContext() as InternalContext<
      Partial<TPCustomer>
    >;

    internalCtx.params = { id: params.id, ...body };
    internalCtx.options!.canCatchError = true;

    const result = await pcustomerDao.updatePCustomer(internalCtx);

    return result;
  } catch (error) {
    if (isStandardError(error)) return error;

    const err = new AppError("Cannot update existing potential customer");
    err.asHTTPError("InternalServerError");
    return err;
  }
}

File: python/src/core/modules/pcustomer_management/functions/update_customer.py
# Import from core
from core.context.runtime_context import RuntimeContext
from core.context.internal_context.main import initialize_internal_context
from core.error import AppError, is_standard_error
from core.modules.pcustomer_management.data_model.dao import PCustomerDAO


async def update_customer(ctx: RuntimeContext):
    """
    Cập nhật thông tin của một khách hàng trong hệ thống.

    Args:
        ctx (RuntimeContext): Ngữ cảnh runtime chứa thông tin request.

    Returns:
        dict | AppError: Thông tin khách hàng vừa cập nhật hoặc lỗi chuẩn hóa.
    """
    try:
        body = await ctx.get_body()
        params = ctx.get_params()

        pcustomer_dao = PCustomerDAO()
        internal_ctx = initialize_internal_context()
        internal_ctx.params = {"id": params.get("id"), **body}
        internal_ctx.options["can_catch_error"] = True

        result = await pcustomer_dao.update_pcustomer(internal_ctx)
        return result

    except Exception as error:
        if is_standard_error(error):
            return error

        err = AppError("Cannot update existing potential customer")
        err.as_http_error("InternalServerError")
        return err
